File: lib/src/models/time_modification.dart
import 'package:meta/meta.dart';
import 'package:musicxml_parser/src/exceptions/musicxml_validation_exception.dart';

/// Represents a time modification, such as a tuplet.
@immutable
class TimeModification {
  /// The number of actual notes in the tuplet (e.g., 3 for a triplet).
  final int actualNotes;

  /// The number of normal notes of the same type that would normally
  /// occupy the same duration (e.g., 2 for a triplet of eighths).
  final int normalNotes;

  /// The note type of the normal notes (e.g., "eighth", "quarter").
  /// Optional in MusicXML.
  final String? normalType;

  /// The number of dots on the normal notes, if specified.
  /// This represents the count of <normal-dot/> elements.
  /// Null if <normal-dot> elements are not present.
  final int? normalDotCount;

  /// Creates a new [TimeModification] instance.
  const TimeModification({
    required this.actualNotes,
    required this.normalNotes,
    this.normalType,
    this.normalDotCount,
  });

  /// Creates a new [TimeModification] instance with validation.
  factory TimeModification.validated({
    required int actualNotes,
    required int normalNotes,
    String? normalType,
    int? normalDotCount,
    int? line, // For error reporting
    Map<String, dynamic>? context, // For error reporting
  }) {
    if (actualNotes <= 0) {
      throw MusicXmlValidationException(
        'TimeModification actualNotes must be positive, got $actualNotes',
        rule: 'time_modification_actual_notes_positive',
        line: line,
        context: {'actualNotes': actualNotes, ...?context},
      );
    }
    if (normalNotes <= 0) {
      throw MusicXmlValidationException(
        'TimeModification normalNotes must be positive, got $normalNotes',
        rule: 'time_modification_normal_notes_positive',
        line: line,
        context: {'normalNotes': normalNotes, ...?context},
      );
    }
    // normalDotCount can be null (not specified) or >= 0.
    // A count of 0 means <normal-dot> was not present or explicitly zero (though MusicXML implies presence means at least one).
    // The parser will set this to the number of <normal-dot> elements if they exist, otherwise null.
    if (normalDotCount != null && normalDotCount < 0) {
      throw MusicXmlValidationException(
        'TimeModification normalDotCount cannot be negative if specified, got $normalDotCount',
        rule: 'time_modification_normal_dot_count_non_negative',
        line: line,
        context: {'normalDotCount': normalDotCount, ...?context},
      );
    }

    return TimeModification(
      actualNotes: actualNotes,
      normalNotes: normalNotes,
      normalType: normalType,
      normalDotCount: normalDotCount,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TimeModification &&
          runtimeType == other.runtimeType &&
          actualNotes == other.actualNotes &&
          normalNotes == other.normalNotes &&
          normalType == other.normalType &&
          normalDotCount == other.normalDotCount;

  @override
  int get hashCode =>
      actualNotes.hashCode ^
      normalNotes.hashCode ^
      (normalType?.hashCode ?? 0) ^
      (normalDotCount?.hashCode ?? 0); // Handle null for normalDotCount

  @override
  String toString() {
    final parts = [
      'actualNotes: $actualNotes',
      'normalNotes: $normalNotes',
      if (normalType != null) 'normalType: $normalType',
      if (normalDotCount != null) 'normalDotCount: $normalDotCount',
    ];
    return 'TimeModification{${parts.join(', ')}}';
  }
}

File: csharp/MusicXMLParser/Models/TimeModification.cs
using System;
using System.Collections.Generic;
using MusicXMLParser.Exceptions; // Assuming exceptions are in this namespace

namespace MusicXMLParser.Models
{
    /// <summary>
    /// Represents a time modification, such as a tuplet.
    /// </summary>
    public class TimeModification
    {
        /// <summary>
        /// The number of actual notes in the tuplet (e.g., 3 for a triplet).
        /// </summary>
        public int ActualNotes { get; }

        /// <summary>
        /// The number of normal notes of the same type that would normally
        /// occupy the same duration (e.g., 2 for a triplet of eighths).
        /// </summary>
        public int NormalNotes { get; }

        /// <summary>
        /// The note type of the normal notes (e.g., "eighth", "quarter").
        /// Optional in MusicXML.
        /// </summary>
        public string NormalType { get; }

        /// <summary>
        /// The number of dots on the normal notes, if specified.
        /// This represents the count of &lt;normal-dot/&gt; elements.
        /// Null if &lt;normal-dot&gt; elements are not present.
        /// </summary>
        public int? NormalDotCount { get; }

        /// <summary>
        /// Creates a new <see cref="TimeModification"/> instance.
        /// </summary>
        public TimeModification(
            int actualNotes,
            int normalNotes,
            string normalType = null,
            int? normalDotCount = null)
        {
            if (actualNotes <= 0)
                throw new ArgumentOutOfRangeException(nameof(actualNotes), "ActualNotes must be positive.");
            if (normalNotes <= 0)
                throw new ArgumentOutOfRangeException(nameof(normalNotes), "NormalNotes must be positive.");
            if (normalDotCount.HasValue && normalDotCount.Value < 0)
                throw new ArgumentOutOfRangeException(nameof(normalDotCount), "NormalDotCount cannot be negative if specified.");

            ActualNotes = actualNotes;
            NormalNotes = normalNotes;
            NormalType = normalType;
            NormalDotCount = normalDotCount;
        }

        // Removed CreateValidated static method as per user request to defer validation.
        // Constructor with basic argument checks remains.

        public override bool Equals(object obj)
        {
            if (obj is TimeModification other)
            {
                return ActualNotes == other.ActualNotes &&
                       NormalNotes == other.NormalNotes &&
                       NormalType == other.NormalType &&
                       NormalDotCount == other.NormalDotCount;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(ActualNotes, NormalNotes, NormalType, NormalDotCount);
        }

        public override string ToString()
        {
            var parts = new List<string>
            {
                $"ActualNotes: {ActualNotes}",
                $"NormalNotes: {NormalNotes}"
            };
            if (!string.IsNullOrEmpty(NormalType))
            {
                parts.Add($"NormalType: {NormalType}");
            }
            if (NormalDotCount.HasValue)
            {
                parts.Add($"NormalDotCount: {NormalDotCount.Value}");
            }
            return $"TimeModification{{{string.Join(", ", parts)}}}";
        }
    }
}
